src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,488 @@
+            using (var jsonReader = new BsonBinaryReader(stream))
+            {
+                var context = BsonDeserializationContext.CreateRoot(jsonReader);
+                return BsonDocumentSerializer.Instance.Deserialize(context);
+            }
+        }
+
+        private static byte[] ToBytes(BsonDocument doc)
+        {
+            BsonBinaryWriterSettings settings = new BsonBinaryWriterSettings()
+            {
+#pragma warning disable 618
+                GuidRepresentation = GuidRepresentation.Standard
+#pragma warning restore 618
+            };
+            return doc.ToBson(null, settings);
+        }
+
+        private class ClientFirst : ISaslStep
+        {
+            private readonly byte[] _bytesToSendToServer;
+            private readonly byte[] _nonce;
+            private readonly UsernamePasswordCredential _credential;
+            private readonly string _sessionToken;
+
+            public ClientFirst(
+                byte[] bytesToSendToServer,
+                byte[] nonce,
+                UsernamePasswordCredential credential,
+                string sessionToken)
+            {
+                _bytesToSendToServer = bytesToSendToServer;
+                _nonce = nonce;
+                _credential = credential;
+                _sessionToken = sessionToken;
+            }
+
+            public byte[] BytesToSendToServer
+            {
+                get { return _bytesToSendToServer; }
+            }
+
+            public bool IsComplete
+            {
+                get { return false; }
+            }
+
+            public ISaslStep Transition(SaslConversation conversation, byte[] bytesReceivedFromServer)
+            {
+                var serverFirstMessageDoc = ToDocument(bytesReceivedFromServer);
+                var serverNonce = serverFirstMessageDoc["s"].AsByteArray;
+                var host = serverFirstMessageDoc["h"].AsString;
+
+                if (serverNonce.Length != randomLength * 2 || !serverNonce.Take(randomLength).SequenceEqual(_nonce))
+                {
+                    throw new MongoAuthenticationException(conversation.ConnectionId, message: "Server sent an invalid nonce.");
+                }
+
+                var tuple = AwsSignatureVersion4.SignRequest(
+                    _credential.Username,
+                    _credential.GetInsecurePassword(),
+                    _sessionToken,
+                    serverNonce,
+                    host);
+
+                var document = new BsonDocument()
+                    .Add("a", tuple.Item1)
+                    .Add("d", tuple.Item2);
+
+                if (_sessionToken != null)
+                {
+                    document.Add("t", _sessionToken);
+                }
+
+                var clientSecondMessageBytes = ToBytes(document);
+
+                return new ClientLast(clientSecondMessageBytes);
+            }
+        }
+
+        private class ClientLast : ISaslStep
```

Comment:
it looks like this class does nothing (unless I didn't  miss something)? Is it expected?

Reply:
I don't mind either way: I defer to Mikalai here :)

Reply:
I'm ok with the current version